Synchronize with cancellation when removing a task from a task group.

rjmccall opened the pull request:
We were detaching the child by just modifying the list, but the cancellation path was assuming that that would not be done without holding the task status lock.

Fixes rdar://88398824

rjmccall commented:
@swift-ci Please test

al45tair commented on stdlib/public/BackDeployConcurrency/CompatibilityOverrideConcurrency.def:
```diff
@@ -278,6 +278,10 @@ OVERRIDE_TASK_GROUP(taskGroup_addPending, bool,
                     swift::, (TaskGroup *group, bool unconditionally),
                     (group, unconditionally))
 
+OVERRIDE_TASK_GROUP(task_cancel_group_child_tasks, void,
```
Is it safe to re-order these? There's a big warning at the top of the file suggesting that "Moving or removing entries in this file will break the ABI".

al45tair commented on stdlib/public/CompatibilityOverride/CompatibilityOverrideConcurrency.def:
```diff
@@ -282,6 +282,9 @@ OVERRIDE_TASK_GROUP(taskGroup_addPending, bool,
                     swift::, (TaskGroup *group, bool unconditionally),
                     (group, unconditionally))
 
+OVERRIDE_TASK_GROUP(task_cancel_group_child_tasks, void,
```
See earlier comment — this has moved, which the comment at the top of this file suggests is an ABI break.

al45tair commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -585,18 +605,18 @@ static void fillGroupNextResult(TaskFutureWaitAsyncContext *context,
+  assert(completedTask == readyItem.getTask());
+  assert(readyItem.getTask()->isFuture());
+  readyQueue.enqueue(readyItem);
 }
```
(Note for other reviewers) This change isn't needed in the BackDeployConcurrency code because the `enqueueCompletedTask()` method doesn't exist there. I think it's only a comment change and some whitespace anyway.

al45tair commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -607,6 +627,15 @@ void TaskGroupImpl::offer(AsyncTask *completedTask, AsyncContext *context) {
+  // task), we will need to retain it, which we do in enqueueCompletedTask.
+  // This is wasteful, and the task completion function should be fixed to
+  // transfer ownership of a retain into this function, in which case we
+  // will need to release in the other path.
```
We should have a GitHub Issue or a Radar or something for the fix.

al45tair commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -641,37 +670,54 @@ void TaskGroupImpl::offer(AsyncTask *completedTask, AsyncContext *context) {
+        // the waiting task.  This is a little wasteful; there's no reason
+        // we can't just have the parent task set itself up as a waiter.
+        // But since it's what we're doing, we basically take the same
+        // path as we would if there wasn't a waiter.
```
Does this also want a Radar or something to track fixing it so the parent sets itself up as a waiter?

rokhinip replied:
This is a much better comment than what I wrote - thank you.

al45tair commented on stdlib/public/BackDeployConcurrency/TaskGroup.cpp:
```diff
@@ -716,9 +757,11 @@ static void swift_taskGroup_wait_next_throwingImpl(
+      // Remove the child from the task group's running tasks list.
+      _swift_taskGroup_detachChild(asAbstract(group), completedTask);
+
+      // Balance the retain done by enqueueCompletedTask.
```
Should be `completedTask` here?

rokhinip replied:
No I think this is in reference to the `enqueueCompletedTask` function.

rjmccall replied:
Yeah.  Which doesn't exist in the BackDeployConcurrency library, so I should fix the comment.

DougGregor commented:
Rather than change `BackDeployConcurrency`, we should update `CompatibilityConcurrency56` so this fix will also address problems with the already-shipped Swift 5.6 runtimes in addition to the back-deployed dylibs.

rjmccall commented:
> Rather than change `BackDeployConcurrency`, we should update `CompatibilityConcurrency56` so this fix will also address problems with the already-shipped Swift 5.6 runtimes in addition to the back-deployed dylibs.

Ah, I seem to have completely misunderstood which library was which; this is my first time needing a patch.  `BackDeployConcurrency` is the pre-5.6-only library while `CompatibilityConcurrency56` is the patch library?

rokhinip commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -504,18 +505,37 @@ static void swift_taskGroup_initializeImpl(TaskGroup *group, const Metadata *T)
-  // The counterpart of this (detachChild) is performed by the group itself,
-  // when it offers the completed (child) task's value to a waiting task -
-  // during the implementation of `await group.next()`.
+  // Add the child task to this task group.  The corresponding removal
```
I thought we had discussed offline about changing the behaviour so that the child tasks won't remove themselves from the list even after they complete (ie no modification in `offer`) and the parent would do it instead? This would make sure that we don't have contention on both the TaskGroup atomic and the parent's task status record lock?

rokhinip replied:
If that's a follow-up change and this is more targeted, that's totally fine. Just wanted to understand the plan.

ktoso replied:
Hmm, I guess that's a tradeoff between memory use of tasks hanging around even though they're completed then, wouldn't it be? Also just thinking out loud here though. Optimization wise it'd be great to revisit and get rid of the lock in the group impl, the status acquire/releases should do enough to guarantee synchronization but I never go to follow through on it :(

rokhinip replied:
> Hmm, I guess that's a tradeoff between memory use of tasks hanging around even though they're completed then, wouldn't it be? 

Don't you already have to pay this cost today and keep the children around anyways, since only the first child task which unblocks the parent, can fill the result directly into the parent task? All other child tasks which may have to wait for the parent to consume them, have to hang around anyways. 

> Optimization wise it'd I'd try to revisit and get rid of the lock in the group impl, the status acquire/releases should do enough to guarantee synchronization but I never go to follow through on it :(

I'm not talking about the TaskGroup lock here though? I'm talking about the fact that in this case, we have contention on the TaskGroup atomic as well as on the parent task's task status record lock.

rjmccall replied:
@rokhinip The current patch only synchronously removes tasks from the active-tasks list during `offer` if the parent task is currently waiting, so I don't think this should increase contention on the task status atomic unless there's a contending cancelling thread or something.

The memory for a task group child sticks around until the parent awaits that task, so there isn't any difference in memory use here.  Currently that is required because the parent needs the result that's stored in the child's future, which is allocated inline in the task.  That would not be necessary if we separately allocated the future; we'd be able to immediately throw away the rest of the memory.  We could use a more efficient future implementation with task group children anyway because (1) there's only one possible waiter and (2) it's mediated through the task group rather than directly through the task handle.

rokhinip commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -941,19 +994,45 @@ static void swift_taskGroup_cancelAllImpl(TaskGroup *group) {
 
+SWIFT_CC(swift)
+static void swift_task_cancel_group_child_tasksImpl(TaskGroup *group) {
+  // TaskGroup is not a Sendable type, and so this operation (which is not
```
We could also have an assert that current task == task group's owning task.

rokhinip commented:
Really appreciate the extensive documentation - thank you!

DougGregor commented:
> Ah, I seem to have completely misunderstood which library was which; this is my first time needing a patch. `BackDeployConcurrency` is the pre-5.6-only library while `CompatibilityConcurrency56` is the patch library?

That's correct; `CompatibilityConcurrency56` is the patch library and the only one we should update.

ktoso commented on include/swift/ABI/TaskGroup.h:
```diff
@@ -44,10 +44,14 @@ class alignas(Alignment_TaskGroup) TaskGroup {
 
+  // Remove a child task from the task group. Always called while holding
+  // the status record lock of the task group's owning task.
+  void removeChildTask(AsyncTask *task);
```
Great, thank you 👍

ktoso commented on stdlib/public/BackDeployConcurrency/TaskStatus.cpp:
```diff
@@ -430,14 +430,27 @@ static void swift_taskGroup_attachChildImpl(TaskGroup *group,
   // Acquire the status record lock of parent - we want to synchronize with
   // concurrent cancellation or escalation as we're adding new tasks to the
   // group.
+  auto parent = child->childFragment()->getParent();
```
oh great cleanup to avoid that lookup, thanks

ktoso reviewed (approved):
Thank you John, great cleanup and also the added documentation really helps. Thanks!

rjmccall committed:
Synchronize with cancellation when removing a task from a task group

We were detaching the child by just modifying the list, but the cancellation path was assuming that that would not be done without holding the task status lock.

This patch just fixes the current runtime; the back-deployment side is complicated.

Fixes rdar://88398824

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please smoke test

etcwilde reviewed (commented):
Confirming that the compatibility library comments look accurate.

rjmccall commented:
@swift-ci Please smoke test macOS
